packages/react-components/react-teaching-popover-preview/src/components/TeachingPopoverCarousel/Carousel/Carousel.tsx

```diff
@@ -0,0 +1,145 @@
+import {
+  isHTMLElement,
+  useMergedRefs,
+  useControllableState,
+  type EventHandler,
+  useEventCallback,
+} from '@fluentui/react-utilities';
+import * as React from 'react';
+
+import { CAROUSEL_ITEM } from './constants';
+import { useCarouselWalker_unstable } from './useCarouselWalker';
+import { createCarouselStore } from './createCarouselStore';
+import type { CarouselPageChangeData } from './Carousel.types';
+import { CarouselContextValue } from './CarouselContext';
+
+export type UseCarouselOptions = {
+  defaultValue?: string;
+  value?: string;
+
+  onPageChange?: EventHandler<CarouselPageChangeData>;
+  onFinish?: EventHandler<CarouselPageChangeData>;
+};
+
+// TODO: Migrate this into an external @fluentui/carousel component
+// For now, we won't export this publicly, is only for internal TeachingPopover use until stabilized.
+export function useCarousel_unstable(options: UseCarouselOptions) {
+  const { onPageChange, onFinish } = options;
+
+  const { ref: carouselRef, walker: carouselWalker } = useCarouselWalker_unstable();
+  const [store] = React.useState(() => createCarouselStore());
+
+  const [value, setValue] = useControllableState({
+    defaultState: options.defaultValue,
+    state: options.value,
+    initialState: null,
+  });
+  const rootRef = React.useRef<HTMLDivElement>(null);
+
+  if (process.env.NODE_ENV !== 'production') {
+    // eslint-disable-next-line react-hooks/rules-of-hooks
+    React.useEffect(() => {
+      if (value === null) {
+        // eslint-disable-next-line no-console
+        console.error(
+          'useCarousel: Carousel needs to have a `defaultValue` or `value` prop set. If you want to control the value, use the `value` prop.',
+        );
+      }
+    }, [value]);
+  }
+
+  React.useEffect(() => {
+    const allItems = rootRef.current?.querySelectorAll(`[${CAROUSEL_ITEM}]`)!;
+
+    for (let i = 0; i < allItems.length; i++) {
+      store.addValue(allItems.item(i).getAttribute(CAROUSEL_ITEM)!);
+    }
+
+    return () => {
+      store.clear();
+    };
+  }, [store]);
+
+  React.useEffect(() => {
+    const config: MutationObserverInit = {
+      attributes: true,
+      attributeFilter: [CAROUSEL_ITEM],
+      childList: true,
+      subtree: true,
+    };
+
+    // Callback function to execute when mutations are observed
+    const callback: MutationCallback = mutationList => {
+      for (const mutation of mutationList) {
+        for (const addedNode of Array.from(mutation.addedNodes)) {
+          if (isHTMLElement(addedNode) && addedNode.hasAttribute(CAROUSEL_ITEM)) {
+            const newValue = addedNode.getAttribute(CAROUSEL_ITEM)!;
+            let previousNode = addedNode.previousElementSibling;
```

Comment:
you should use the carouselWalker here instead of `previousElementSibling` because it's possible users might want to add additional HTML wrappers. It doesn't make sense to do so, but you never know what users will do 🤦